### ISSUE TITLE
Interface for custom covariance estimation algorithms

### ISSUE THREAD

**Issue by mateuszbaran** [JuliaStats/StatsBase.jl#439]
Here: https://github.com/JuliaStats/MultivariateStats.jl/pull/84#discussion_r244261509 an issue was raised that it would be a good idea to have an abstract interface for custom covariance estimation methods either in StatsBase or Statistics stdlib. I've recently been working on https://github.com/mateuszbaran/CovarianceEstimation.jl that provides such interface. I could make a PR here but I'm not sure what related methods of `cov` (if any) should be added to StatsBase.

**Comment by nalimilan** [comment]
Sure. What would be needed concretely? Just `abstract type AbstractCovarianceEstimator end`?

Cc: @gragusa

**Comment by nalimilan** [comment]
We can add an empty method for `::AbstractCovarianceEstimator`, with docstrings showing the API, like we do for statistical models. But defining a `SimpleCovarianceEstimator` in StatsBase doesn't sound very useful if we don't provide other types.

**Comment by mateuszbaran** [comment]
What do you mean by an empty method for `::AbstractCovarianceEstimator`? API description could just be in the docstring for `AbstractCovarianceEstimator`.

As for `SimpleCovarianceEstimator`, I was thinking about it and I agree that it should remain in CovarianceEstimation.jl. API description in a docstring is sufficient.

**Comment by mateuszbaran** [comment]
OK, so adding something like this:
```julia

abstract type AbstractCovarianceEstimator end

"""
    cov(x::AbstractVector, ce::AbstractCovarianceEstimator)

Return the sample variance of the vector `x` using estimator `ce`.
"""
cov(x::AbstractVector, ce::AbstractCovarianceEstimator) =
    error("cov is not defined for $(typeof(x)) and $(typeof(ce))")


"""
    cov(x::AbstractVector, y::AbstractVector, ce::AbstractCovarianceEstimator)

Return the covariance of the vectors `x` and `y` using estimator `ce`.
"""
cov(x::AbstractVector, y::AbstractVector, ce::AbstractCovarianceEstimator) =
    error("cov is not defined for $(typeof(x)), $(typeof(y)) and $(typeof(ce))")


"""
    cov(X::AbstractMatrix, ce::AbstractCovarianceEstimator, vardim::Int=1)

Return the covariance matrix associated with `X` along dimension `vardim`
using estimator `ce`.
"""
cov(X::AbstractMatrix, ce::AbstractCovarianceEstimator, vardim::Int=1) =
    error("cov is not defined for $(typeof(X)) and $(typeof(ce))")
```
to `cov.jl` would be correct, right? I'll prepare a PR.

**Comment by ararslan** [comment]
Here's my two cents...

Something that [Jackknife](https://github.com/ararslan/Jackknife.jl) does is that it defines an `Estimator` type that subtypes `Function` and wraps a function, then enforces when the function is called that it returns the expected type (which in Jackknife's case is `Real` or `Missing`). We could take a similar approach here and define something along the lines of
```julia
abstract type AbstractEstimator <: Function end
abstract type CovarianceEstimator <: AbstractEstimator end
cov(est::CovarianceEstimator, x::AbstractMatrix) = ...
```
Jackknife also permits the use of regular ol' functions as estimators, in which case we would similarly define
```julia
cov(f::Function, x::AbstractMatrix) = cov(DefaultEstimator(f), x)
```
for whatever `DefaultEstimator` is.

Also note that in the method signatures, the function/estimator should go first, as is conventional in Julia, but is also in keeping with GLM's use of the model type first, to which this is vaguely akin.

**Comment by mateuszbaran** [comment]
So, something like this?
```julia
abstract type AbstractEstimator <: Function end
abstract type CovarianceEstimator <: AbstractEstimator end

"""
    cov(ce::CovarianceEstimator, x::AbstractVector)

Return the sample variance of the vector `x` using estimator `ce`.
"""
cov(ce::CovarianceEstimator, x::AbstractVector) =
    error("cov is not defined for $(typeof(ce)) and $(typeof(x))")


"""
    cov(ce::CovarianceEstimator, x::AbstractVector, y::AbstractVector)

Return the covariance of the vectors `x` and `y` using estimator `ce`.
"""
cov(ce::CovarianceEstimator, x::AbstractVector, y::AbstractVector) =
    error("cov is not defined for $(typeof(ce)), $(typeof(x)) and $(typeof(y))")


"""
    cov(ce::CovarianceEstimator, X::AbstractMatrix, vardim::Int=1)

Return the covariance matrix associated with `X` along dimension `vardim`
using estimator `ce`.
"""
cov(ce::CovarianceEstimator, X::AbstractMatrix, vardim::Int=1) =
    error("cov is not defined for $(typeof(ce)) and $(typeof(X))")

struct DefaultCovarianceEstimator
    f::Function
end

(ce::DefaultCovarianceEstimator)(x::AbstractVector{T}) where T<:Union{Number,Missing} =
    ce.f(x)::Union{Number, Missing}

(ce::DefaultCovarianceEstimator)(x::AbstractVector{T}, y::AbstractVector{T}) where T<:Union{Number,Missing} =
    ce.f(x, y)::Union{Number, Missing}

(ce::DefaultCovarianceEstimator)(X::AbstractMatrix{T}, vardim::Int=1) where T<:Union{Number,Missing} =
    ce.f(X, vardim)::Union{AbstractMatrix{T}, Missing}

cov(ce::DefaultCovarianceEstimator, x::AbstractVector{T}) where T<:Union{Number,Missing} =
    ce(x)

cov(ce::DefaultCovarianceEstimator, x::AbstractVector{T}, y::AbstractVector{T}) where T<:Union{Number,Missing} =
    ce(x, y)

cov(ce::DefaultCovarianceEstimator, X::AbstractMatrix, vardim::Int=1) =
    ce(X, vardim)

cov(f::Function, x::AbstractVector) = cov(DefaultCovarianceEstimator(f), x)

cov(f::Function, x::AbstractVector, y::AbstractVector) = cov(DefaultCovarianceEstimator(f), x, y)

cov(f::Function, X::AbstractMatrix, vardim::Int=1) = cov(DefaultCovarianceEstimator(f), X, vardim)
```
Do I understand it correctly?

I've written `Number` here because for example `cov` from Statistics also works on complex-valued data.

**Comment by andreasnoack** [comment]
FWIW I'm not in favor of approaches that wraps `Functions`. I generally prefer to define `structs` and overload methods.

**Comment by nalimilan** [comment]
Implemented by #460.
